File: knowledge-graph-app/backend/routers/documents.py
"""Document router — upload, check-duplicate, list, detail, and delete endpoints."""

import logging
from pathlib import Path
from typing import Any

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, UploadFile
from pydantic import BaseModel
from sqlalchemy.orm import Session

from db.session import get_db
from models.models import (
    ActionItem,
    Document,
    Entity,
    EntityDocumentMention,
    NodeDisplay,
    Relationship,
)
from models.schemas import DocumentDetailSchema, DocumentSchema
from routers.deps import CurrentUser
from services.extractor import extract_text
from services.pipeline import run_pipeline
from services.similarity import compute_fingerprint, fingerprint_to_json, find_duplicates

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/documents", tags=["documents"])


# ---------------------------------------------------------------------------
# Duplicate check schemas
# ---------------------------------------------------------------------------


class DuplicateMatch(BaseModel):
    document_id: int
    filename: str
    similarity: float  # 0.0–1.0


class DuplicateCheckResponse(BaseModel):
    has_duplicates: bool
    matches: list[DuplicateMatch]


# ---------------------------------------------------------------------------
# Duplicate check endpoint
# ---------------------------------------------------------------------------


@router.post("/check-duplicate", response_model=DuplicateCheckResponse)
async def check_duplicate(
    file: UploadFile,
    current_user: CurrentUser,
    db: Session = Depends(get_db),
    threshold: float = 0.90,
) -> DuplicateCheckResponse:
    """
    Check whether an uploaded file is highly similar to any existing document.

    The file is NOT stored — this is a read-only pre-upload check.
    Returns a list of existing documents whose similarity to the candidate
    meets or exceeds *threshold* (default 0.90 = 90%).
    """
    data = await file.read()
    filename = file.filename or "upload"

    try:
        candidate_text = extract_text(filename, data)
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc

    # Load only the lightweight columns needed for duplicate detection.
    # raw_text is fetched only for legacy rows that lack a stored fingerprint.
    existing = db.query(
        Document.id,
        Document.filename,
        Document.raw_text,
        Document.fingerprint,
    ).all()
    existing_tuples = [
        (row.id, row.filename, row.raw_text, row.fingerprint)
        for row in existing
    ]

    matches = find_duplicates(candidate_text, existing_tuples, threshold=threshold)

    return DuplicateCheckResponse(
        has_duplicates=len(matches) > 0,
        matches=[DuplicateMatch(**m) for m in matches],
    )


# ---------------------------------------------------------------------------
# Upload endpoint
# ---------------------------------------------------------------------------


@router.post("/upload", response_model=DocumentSchema, status_code=201)
async def upload_document(
    file: UploadFile,
    background_tasks: BackgroundTasks,
    current_user: CurrentUser,
    db: Session = Depends(get_db),
) -> DocumentSchema:
    """Accept a multipart file upload, extract text, persist, and queue pipeline."""
    data = await file.read()
    filename = file.filename or "upload"
    file_type = Path(filename).suffix.lower().lstrip(".") or "unknown"

    try:
        raw_text = extract_text(filename, data)
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc

    doc = Document(
        uploader_user_id=current_user["id"],
        filename=filename,
        raw_text=raw_text,
        file_type=file_type,
        processed_at=None,
        fingerprint=fingerprint_to_json(compute_fingerprint(raw_text)),
    )
    db.add(doc)
    db.commit()
    db.refresh(doc)

    background_tasks.add_task(run_pipeline, doc.id, db)

    return DocumentSchema.model_validate(doc)


@router.get("", response_model=list[DocumentSchema])
def list_documents(
    current_user: CurrentUser,
    db: Session = Depends(get_db),
) -> list[DocumentSchema]:
    """Return all documents uploaded by the current user."""
    docs = (
        db.query(Document)
        .filter(Document.uploader_user_id == current_user["id"])
        .order_by(Document.created_at.desc())
        .all()
    )
    return [DocumentSchema.model_validate(d) for d in docs]


@router.get("/{doc_id}", response_model=DocumentDetailSchema)
def get_document(
    doc_id: int,
    current_user: CurrentUser,
    db: Session = Depends(get_db),
) -> DocumentDetailSchema:
    """Return document metadata and raw text; 404 if not found or not owned by caller."""
    doc = (
        db.query(Document)
        .filter(Document.id == doc_id, Document.uploader_user_id == current_user["id"])
        .first()
    )
    if doc is None:
        raise HTTPException(status_code=404, detail="Document not found.")
    return DocumentDetailSchema.model_validate(doc)


@router.delete("/{doc_id}", status_code=204, response_model=None)
def delete_document(
    doc_id: int,
    current_user: CurrentUser,
    db: Session = Depends(get_db),
) -> None:
    """
    Delete a document and clean up all data derived from it.

    Cascade order:
      1. Delete action items belonging to the document.
      2. Delete entity_document_mentions for the document.
      3. Delete entities that are now mentioned in NO other document
         (orphan entities), along with their relationships and node_display rows.
      4. Delete the document itself.
    """
    doc = (
        db.query(Document)
        .filter(Document.id == doc_id, Document.uploader_user_id == current_user["id"])
        .first()
    )
    if doc is None:
        raise HTTPException(status_code=404, detail="Document not found.")

    # 1 — Remove action items
    db.query(ActionItem).filter(ActionItem.document_id == doc_id).delete(
        synchronize_session=False
    )

    # 2 — Collect entities mentioned in this document before we remove mentions
    mentioned_entity_ids = [
        row.entity_id
        for row in db.query(EntityDocumentMention.entity_id)
        .filter(EntityDocumentMention.document_id == doc_id)
        .all()
    ]

    # Remove mentions for this document
    db.query(EntityDocumentMention).filter(
        EntityDocumentMention.document_id == doc_id
    ).delete(synchronize_session=False)

    # 3 — Find orphan entities (no remaining mentions in any other document)
    orphan_ids = [
        eid
        for eid in mentioned_entity_ids
        if db.query(EntityDocumentMention)
        .filter(EntityDocumentMention.entity_id == eid)
        .count()
        == 0
    ]

    if orphan_ids:
        # Remove node_display cache rows for orphans
        db.query(NodeDisplay).filter(NodeDisplay.entity_id.in_(orphan_ids)).delete(
            synchronize_session=False
        )
        # Remove relationships where either endpoint is an orphan
        db.query(Relationship).filter(
            Relationship.entity_a_id.in_(orphan_ids)
            | Relationship.entity_b_id.in_(orphan_ids)
        ).delete(synchronize_session=False)
        # Remove the orphan entities themselves
        db.query(Entity).filter(Entity.id.in_(orphan_ids)).delete(
            synchronize_session=False
        )

    # 4 — Remove the document
    db.delete(doc)
    db.commit()

    logger.info(
        "Deleted document %d (%s); removed %d orphan entities.",
        doc_id,
        doc.filename,
        len(orphan_ids),
    )

File: knowledge-graph-app/frontend/src/api/documents.ts
import { getAuthHeader } from "./auth";

export interface DuplicateMatch {
  document_id: number;
  filename: string;
  similarity: number; // 0.0–1.0
}

export interface DuplicateCheckResponse {
  has_duplicates: boolean;
  matches: DuplicateMatch[];
}

export interface DocumentRecord {
  id: number;
  uploader_user_id: number;
  filename: string;              // AI-generated descriptive name (after pipeline)
  original_filename: string | null;  // user-supplied name at upload time
  file_type: string;
  created_at: string;
  processed_at: string | null;
  ai_category: string | null;
}

export interface DocumentDetailRecord extends DocumentRecord {
  raw_text: string;
}

export async function uploadDocument(
  file: File,
  overrideName?: string
): Promise<DocumentRecord> {
  const form = new FormData();
  // If a rename was provided, create a new File with the override name so
  // the multipart field carries the correct filename to the server.
  const uploadFile =
    overrideName && overrideName.trim() && overrideName !== file.name
      ? new File([file], overrideName.trim(), { type: file.type })
      : file;
  form.append("file", uploadFile);
  const res = await fetch("/api/documents/upload", {
    method: "POST",
    headers: getAuthHeader(),
    body: form,
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Upload failed (${res.status}): ${text}`);
  }
  return res.json();
}

export async function listDocuments(): Promise<DocumentRecord[]> {
  const res = await fetch("/api/documents", {
    headers: getAuthHeader(),
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`List documents failed (${res.status}): ${text}`);
  }
  return res.json();
}

export async function getDocument(id: number): Promise<DocumentDetailRecord> {
  const res = await fetch(`/api/documents/${id}`, {
    headers: getAuthHeader(),
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Get document failed (${res.status}): ${text}`);
  }
  return res.json();
}

export async function checkDuplicate(
  file: File
): Promise<DuplicateCheckResponse> {
  const form = new FormData();
  form.append("file", file);
  const res = await fetch("/api/documents/check-duplicate", {
    method: "POST",
    headers: getAuthHeader(),
    body: form,
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Duplicate check failed (${res.status}): ${text}`);
  }
  return res.json();
}

export async function deleteDocument(id: number): Promise<void> {
  const res = await fetch(`/api/documents/${id}`, {
    method: "DELETE",
    headers: getAuthHeader(),
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Delete failed (${res.status}): ${text}`);
  }
}
